trpcultivate_phenotypes/src/Plugin/Validators/ValidDataFile.php

```diff
@@ -0,0 +1,198 @@
+      $file_object = File::load($fid);
+    }
+    elseif ($filename) {
+      // The file input is a string value, a path to the file. Locate the file entity
+      // by uri and load the file object using the returned file id number that matched.
+      $file_entities = $this->service_EntityTypeManager
+        ->getStorage('file')
+        ->loadByProperties(['uri' => $filename]);
+
+      $file_entity = reset($file_entities);
+
+      if ($file_entity) {
+        $fid = $file_entity->get('fid')->value;
+        $file_object = File::load($fid);
+      }
+    }
+
+    // Check that the file input provided returned a file object.
+    if (is_null($file_object)) {
+      return [
+        'case' => 'Filename or file id failed to load a file object',
+        'valid' => FALSE,
+        'failedItems' => ['filename' => $filename, 'fid' => $fid]
+      ];
+    }
+    
+    // File object has loaded successfully. Any subsequent failed test from this point
+    // will reference the filename and file id from the established file object.
+    $file_filename = $file_object->getFileName();
+    $file_fid = $file_object->id();
+
+    // Check that the file is not blank by inspecting the file size to see if it is greater than 0.
+    $file_size = $file_object->getSize();
+    if (!$file_size) {
+      return [
+        'case' => 'The file has no data and is an empty file',
+        'valid' => FALSE,
+        'failedItems' => $failed_items = ['filename' => $file_filename, 'fid' => $file_fid]
+      ];
+    }
+    
+    // Check that both the file MIME type and file extension are supported.
+    $file_mime_type = $file_object->getMimeType();
+    $file_extension = pathinfo($file_filename, PATHINFO_EXTENSION);
+    
+    // Reference supported MIME types and file extensions values set by the importer instance.
+    $supported_file_extensions = $this->getSupportedFileExtensions();
+    $supported_mime_types = $this->getSupportedMimeTypes();
+
+    if (!in_array($file_mime_type, $supported_mime_types)) {
+      if (in_array($file_extension, $supported_file_extensions)) {
+        // The file extension is supported but the MIME type is not.
+        return [
+          'case' => 'Unsupported file MIME type',
+          'valid' => FALSE,
+          'failedItems' => ['mime' => $file_mime_type, 'extension' => $file_extension]
+        ];
+      }
+      else {
+        // Both MIME type and file extension are not supported.
+        return [
+          'case' => 'Unsupported file mime type and mismatched extension',
+          'valid' => FALSE,
+          'failedItems' => ['mime' => $file_mime_type, 'extension' => $file_extension]
+        ];
+      }
+    }
+
+    // Check that the file can be opened.
+    $file_uri  = $file_object->getFileUri();
+    $file_handle = @fopen($file_uri, 'r');
+    
+    if (!$file_handle) {
+      return [
+        'case' => 'The file cannot be opened',
```

Comment:
To be consistent with line 193:
```suggestion
        'case' => 'Data file cannot be opened',
```